### DISCUSSION TITLE
Category Updation for Interactive Video

### DISCUSSION THREAD

**Discussion by Xitija** [Sunbird-inQuiry/Community#15] [Q&A]
Moved this discussion from [here](https://github.com/Sunbird-Knowlg/Community/discussions/15#discussioncomment-2902977)

Hello All,

We have created a new category for Interactive Video and we want it to be configured such that it is selected as a default category in while creating the Interactive Video (Video + Question set) in the portal .
We have the following approaches for doing it:

By updating the channel where the category is picked from for the Question Set
OR
By having a config such that we can have the category listed
Can you please help us in choosing which approach we should go with ?
@coolbung @baljeetsingh @vaivk369 @pallakartheekreddy @Krishnaj20 @vinukumar-vs

We tried following https://knowlg.sunbird.org/learn/product-and-developer-guide/editors but could not get much help.

Attaching a screenshot of the form from https://dock.sunbirded.org/contribute where category is to be updated in the 'Type' form field.
![image](https://user-images.githubusercontent.com/65657373/172588973-281ffc98-df65-4e08-9beb-177baabfd290.png)

**Comment by pallakartheekreddy** [comment]
@Xitija
Can you please provide more details about your options like, 
1. What you will update in the channel where the category is picked from for the Question Set?
2. What config changes are you suggesting? Can you please provide some sample?

Please check the below option also possible.
 Current implementation when the user is trying to create a questionSet in the Interactive Video system will get the list of questionSet primaryCategories at channel level and by default, it will select the first primaryCategory and creates the questionset. 
Here, We can enhance the code instead of selecting the first one we will show the list of questionSet primary categories available at the channel level in a dropdown user can choose one of the primaryCategory and create a question set with that.

**Reply by Xitija** [reply]
@vaivk369 can you please guide us here.

**Reply by Xitija** [reply]
Hello @pallakartheekreddy ,
Currently the category is picked from an array and its first item is picked as the default category . So to solve this we are introducing a system setting where we will be storing the category identifier and that will be picked from a config instead of the first item in the array.

**Reply by pallakartheekreddy** [reply]
@Xitija We should not use category identifier because it will be different if any tenant(channel) overrides the category at the tenant level instead we can use the category name & targetObjectType combination.

**Reply by pallakartheekreddy** [reply]
@Xitija Just checking if you need any assistance on this or is the issue resolved?

**Reply by Xitija** [reply]
@pallakartheekreddy Sir ,
Thanks a lot for your concern .

Currently I am trying to implement the category by setting an environment variable.
I am making changes to the files 'environmentVariablesHelpers.js , clientRoutes.js , index.html' ,
In this way I am fetching it the required file using getElementById method.

**Reply by pallakartheekreddy** [reply]
@Xitija Yes, This is the way we will access env variables in the portal. 
In which release you are targeting this feature?  Hope you were aware of coKreate release dates?

**Reply by Xitija** [reply]
@pallakartheekreddy Sir ,
We are planning to release it in 4.10.

**Comment by surendrasinghs** [comment]
@Xitija Can you help me understand the complete user flow? Here's my understanding so far...

1. Contributor/Creator selects any primary category to start creating. Let's say, I want to create "Explanation". 
2. Contributor uploads a video. As soon as the video is uploaded, Contributor will see the "Make it interactive" button.
3. On clicking "Make it interactive", the upload editor opens a question set editor with pre-filled values such as Name, Primary category, etc. Can you please provide me with the PRD/Form configuration for this?

@karn-nahata

**Reply by Xitija** [reply]
@surendrasinghs Sir,
Here are the detailed steps on how it is implemented.
This is the flow that is used please let me know if you need more details.

1. Contributor goes to the respective project , then goes to upload content , clicks on 'Create New' where user selects type pf content to be uploaded , here we select 'Explaination Content' .
2. After that a video uploader popup is prompted and after a video file is selected user is taken to a Interactive video creation screen.
3. Video preview is shown and a button named 'Add new Question Set' is shown. 
4. On clicking that button a popup for timestamp selection is prompted where the Question set will be added.
5. On choosing the timestamp the user will be presented with a form for creating Question Sets and the questions may be of type Subjective or MCQ.
6. On saving the Question Set a marker is generated at the respective time stamp and the Question Set is loded keeping the video in the background paused.

Here's a video for understanding the creation of Interactive Video , its a bit long , but you can watch it if you find time.
Video >https://drive.google.com/file/d/1SUC18myQu0FIhpR7h3hLFhJvbLPpUA-7/view

**Reply by surendrasinghs** [reply]
@Xitija this is helpful, thanks!

In the video, when the user (contributor) clicks on "Add Question Set"
1. Given/Assume the video is already being played and the timestamp is not 00:00 or the end of the video, will the system automatically capture the timestamp? It should as that would make it easy to insert questions at various timestamps in the video.
2. On the Question Set creation page, the primary category was selected as "Demo Practice Question Set". Where is that configuration stored as of today in your current implementation?
3. What's the expected behaviour regarding the primary category of the Question set? I would assume we want to provide a default primary category for the question set and do not want the contributor to change it.
4. What are we calling this primary category for the question set? I recommend we call it "Interaction Video Questions" (object type: Question Set)
5. Can you share the Interactive Video category that you mentioned in your initial post? Refer to "We have created a new category for Interactive Video". Please create a file in [this folder](https://github.com/Sunbird-Ed/creation-portal/tree/release-4.10.0/reference-sample-configuration) and raise a PR.

Look forward to seeing the proposed primary category definition.

**Reply by Xitija** [reply]
@surendrasinghs Sir ,
Answering your questions in respective order

1. Yes the system automatically captures the timestamp on demand .
2. We have introduced a new category for this type of videos called as 'Interactive Content' . We have configured it as an environment variable in the system.
3. The category 'Interactive Content' will be set as the default category it will be selected by the system and the contributor cannot change it.
4. As mentioned above we are calling it as 'Interactive content' and its object type: QuestionSet
5. What is this file supposed to have ?

**Reply by surendrasinghs** [reply]
I suggest we call it "Interactive Video Question Set". The file is the primary category definition. Please refer to the samples in the folder.

Share a quick draft and we can iterate over it to complete the "Interactive Video Question Set" primary category definition.

**Reply by Xitija** [reply]
@surendrasinghs Sir,
Apologies for the Late Reply;
I am looking into this and I will get back to you as soon as possible.

**Reply by Xitija** [reply]
@surendrasinghs Sir,
I have created the new category with name "Interactive Video Question Set" and added definition to it.
I am sharing details of it in this [document](https://docs.google.com/document/d/1wx5kEVHKvoMJemuo8anZVN3-P0i__nJEboPspqdkr-g/edit?usp=sharing)
Please let me know if anything else is required.
